pkg/storageops/azure/azure.go

```diff
@@ -0,0 +1,808 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return NewAzureClient(instance, subscriptionID, resourceGroupName)
+}
+
+func NewAzureClient(
+	instance, subscriptionID, resourceGroupName string,
+) (storageops.Ops, error) {
+	envName := os.Getenv(EnvAzureEnvironment)
+	if len(envName) == 0 {
+		envName = defaultEnvironment
+	}
+	env, err := azure.EnvironmentFromName(envName)
+	if err != nil {
+		return nil, fmt.Errorf("invalid cloud name '%s' specified: %v", envName, err)
+	}
+
+	tenantID, err := storageops.GetEnvValueStrict(EnvTenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	clientID, err := storageops.GetEnvValueStrict(EnvClientID)
+	if err != nil {
+		return nil, err
+	}
+
+	clientSecret, err := storageops.GetEnvValueStrict(EnvClientSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	oauthConfig, err := adal.NewOAuthConfig(env.ActiveDirectoryEndpoint, tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := adal.NewServicePrincipalToken(*oauthConfig, clientID,
+		clientSecret, env.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	authorizer := autorest.NewBearerAuthorizer(token)
+
+	disksClient := compute.NewDisksClient(subscriptionID)
+	disksClient.Authorizer = authorizer
+	disksClient.PollingDelay = clientPollingDelay
+	disksClient.RetryAttempts = clientRetryAttempts
+	disksClient.AddToUserAgent(userAgentExtension)
+
+	vmsClient := compute.NewVirtualMachinesClient(subscriptionID)
+	vmsClient.Authorizer = authorizer
+	vmsClient.PollingDelay = clientPollingDelay
+	vmsClient.RetryAttempts = clientRetryAttempts
+	vmsClient.AddToUserAgent(userAgentExtension)
+
+	snapshotsClient := compute.NewSnapshotsClient(subscriptionID)
+	snapshotsClient.Authorizer = authorizer
+	snapshotsClient.PollingDelay = clientPollingDelay
+	snapshotsClient.RetryAttempts = clientRetryAttempts
+	snapshotsClient.AddToUserAgent(userAgentExtension)
+
+	return &azureOps{
+		instance:          instance,
+		resourceGroupName: resourceGroupName,
+		disksClient:       &disksClient,
+		vmsClient:         &vmsClient,
+		snapshotsClient:   &snapshotsClient,
+	}, nil
+}
+
+func (a *azureOps) Create(
+	template interface{},
+	labels map[string]string,
+) (interface{}, error) {
+	d, ok := template.(*compute.Disk)
+	if !ok {
+		return nil, storageops.NewStorageError(
+			storageops.ErrVolInval,
+			"Invalid volume template given",
+			a.instance,
+		)
+	}
+
+	ctx := context.Background()
+	future, err := a.disksClient.CreateOrUpdate(
+		ctx,
+		a.resourceGroupName,
+		*d.Name,
+		compute.Disk{
+			Location: d.Location,
+			Type:     d.Type,
+			Zones:    d.Zones,
+			Tags:     formatTags(labels),
+			Sku:      d.Sku,
+			DiskProperties: &compute.DiskProperties{
+				CreationData: &compute.CreationData{
+					CreateOption: compute.Empty,
+				},
+				DiskSizeGB: d.DiskProperties.DiskSizeGB,
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create disk: %v", err)
+	}
+
+	t := func() (interface{}, bool, error) {
```

Comment:
This snippet to wait for completion duplicated in all the backgrounded calls. All the future structs implement `azure.Future`. So maybe you can pass in the future variables to a `waitForCompletion`  API and wait for future.Done() to return true.

Reply:
As discussed offline `azure.Future` is a struct and cannot be used to implement a generic `waitForCompletion()`
In the next version of sdk we will be using the `future.WaitForCompletion` which is built.